### PR TITLE
[ISSUE-828][REFACTOR] Migrate memory tracker related configs to ConfigEntry

### LINKED ISSUE #828: [REFACTOR] Migrate memory tracker related configs to ConfigEntry
### Describe the intention you'd like
A clear and concise description of what you want to do.

/cc @who-need-to-know

/assign @who-can-help-you

### PR DESCRIPTION
**Pull request by nafiyAix**
### What changes were proposed in this pull request?
Migrate memory tracker related configs to ConfigEntry

### What are the items that need reviewer attention?
1. In order to keep the consistency with other timeConf format, changing the default value and the parse mechanism of `rss.worker.memory.check.interval` from 10 to 10ms. Do we have a way to keep both the consistency and compatibility?
2. add version("0.2.0") for other configs which miss it. 

### Related issues.
close #828 

### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 1]
```diff
@@ -1641,32 +1680,66 @@ object CelebornConf extends Logging {
     conf.getTimeAsMs("rss.partition.sort.timeout", "220s")
   }
 
-  def partitionSortMaxMemoryRatio(conf: CelebornConf): Double = {
-    conf.getDouble("rss.partition.sort.memory.max.ratio", 0.1)
-  }
+  val PARTITION_SORT_MAX_MEMORY_RATIO: ConfigEntry[Double] =
+    buildConf("celeborn.partition.sort.memory.maxRatio")
```
Comment (AngersZhuuuu) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1684]: ·celeborn.partition.sorter.memory.maxRatio·

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -1641,32 +1680,66 @@ object CelebornConf extends Logging {
...
+      .version("0.2.0")
+      .doubleConf
+      .checkValue(v => v >= 0.0 && v <= 1.0, "should be in [0.0, 1.0].")
+      .createWithDefault(0.95)
 
-  def workerResumeRatio(conf: CelebornConf): Double = {
-    conf.getDouble("rss.resume.memory.ratio", 0.5)
-  }
+  val WORKER_RESUME_RATIO: ConfigEntry[Double] =
+    buildConf("celeborn.resume.memory.ratio")
```
Comment (AngersZhuuuu) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1714]: celeborn.worker.memory.resume.ratio

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -1641,32 +1680,66 @@ object CelebornConf extends Logging {
...
+      .version("0.2.0")
+      .doubleConf
+      .checkValue(v => v >= 0.0 && v <= 1.0, "should be in [0.0, 1.0].")
+      .createWithDefault(0.85)
 
-  def workerPauseRepcaliteRatio(conf: CelebornConf): Double = {
-    conf.getDouble("rss.pause.replicate.memory.ratio", 0.95)
-  }
+  val WORKER_PAUSE_REPLICATE_RATIO: ConfigEntry[Double] =
+    buildConf("celeborn.pause.replicate.memory.ratio")
```
Comment (AngersZhuuuu) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1704]: celeborn.worker.memory.paush.replicate.ratio

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 4]
```diff
@@ -1641,32 +1680,66 @@ object CelebornConf extends Logging {
...
+      .version("0.2.0")
+      .doubleConf
+      .checkValue(v => v >= 0.0 && v <= 1.0, "should be in [0.0, 1.0].")
+      .createWithDefault(0.1)
 
-  def workerPausePushDataRatio(conf: CelebornConf): Double = {
-    conf.getDouble("rss.pause.pushdata.memory.ratio", 0.85)
-  }
+  val WORKER_PAUSE_PUSH_DATA_RATIO: ConfigEntry[Double] =
+    buildConf("celeborn.pause.pushdata.memory.ratio")
```
Comment (AngersZhuuuu) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1694]: cleborn.worker.memory.paush.pushdata.ratio

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 5]
```diff
@@ -1641,32 +1680,66 @@ object CelebornConf extends Logging {
     conf.getTimeAsMs("rss.partition.sort.timeout", "220s")
   }
 
-  def partitionSortMaxMemoryRatio(conf: CelebornConf): Double = {
-    conf.getDouble("rss.partition.sort.memory.max.ratio", 0.1)
-  }
+  val PARTITION_SORTER_MAX_MEMORY_RATIO: ConfigEntry[Double] =
+    buildConf("celeborn.partition.sorter.memory.maxRatio")
```
Comment (AngersZhuuuu) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1684]: `celeborn.partitionSorter.memory.maxRatio`? cc @pan3793 @waitinfuture

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 6]
```diff
@@ -1641,32 +1680,66 @@ object CelebornConf extends Logging {
...
+      .version("0.2.0")
+      .doubleConf
+      .checkValue(v => v >= 0.0 && v <= 1.0, "should be in [0.0, 1.0].")
+      .createWithDefault(0.1)
 
-  def workerPausePushDataRatio(conf: CelebornConf): Double = {
-    conf.getDouble("rss.pause.pushdata.memory.ratio", 0.85)
-  }
+  val WORKER_PAUSE_PUSH_DATA_RATIO: ConfigEntry[Double] =
+    buildConf("celeborn.worker.memory.pause.pushData.ratio")
```
Comment (AngersZhuuuu) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1694]: `celeborn.worker.pausePushData.memory.ratio` @pan3793

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 7]
```diff
@@ -853,6 +864,7 @@ object CelebornConf extends Logging {
       .doc("Push buffer queue size for a task. The maximum memory is " +
         "`celeborn.push.buffer.max.size` * `celeborn.push.queue.capacity`, " +
         "default: 64KiB * 512 = 32MiB")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:867]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 8]
```diff
@@ -842,6 +852,7 @@ object CelebornConf extends Logging {
         "consideration keep this buffer size higher than 32K. Example: If reducer amount is " +
         "2000, buffer size is 64K, then each task will consume up to `64KiB * 2000 = 125MiB` " +
         "heap memory.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:855]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 9]
```diff
@@ -864,6 +876,7 @@ object CelebornConf extends Logging {
       .doc("Amount of Netty in-flight requests. The maximum memory is " +
         "`celeborn.push.maxReqsInFlight` * `celeborn.push.buffer.max.size` * " +
         "compression ratio(1 in worst case), default: 64Kib * 32 = 2Mib")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:879]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 10]
```diff
@@ -873,6 +886,7 @@ object CelebornConf extends Logging {
       .categories("client")
       .version("0.2.0")
       .doc("Timeout for a task to fetch chunk.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:889]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 11]
```diff
@@ -882,6 +896,7 @@ object CelebornConf extends Logging {
       .categories("client")
       .version("0.2.0")
       .doc("Amount of in-flight chunk fetch request.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:899]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 12]
```diff
@@ -891,6 +906,7 @@ object CelebornConf extends Logging {
       .categories("client")
       .version("0.2.0")
       .doc("Max parallelism of client on sending RPC requests.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:909]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 13]
```diff
@@ -900,6 +916,7 @@ object CelebornConf extends Logging {
       .categories("master")
       .version("0.2.0")
       .doc("Application heartbeat timeout.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:919]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 14]
```diff
@@ -909,6 +926,7 @@ object CelebornConf extends Logging {
       .categories("client")
       .version("0.2.0")
       .doc("Interval for client to send heartbeat message to master.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:929]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 15]
```diff
@@ -918,6 +936,7 @@ object CelebornConf extends Logging {
       .categories("client")
       .version("0.2.0")
       .doc("Interval for client to check expired shuffles.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:939]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 16]
```diff
@@ -927,6 +946,7 @@ object CelebornConf extends Logging {
       .categories("client")
       .version("0.2.0")
       .doc("Interval for client to refresh excluded worker list.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:949]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 17]
```diff
@@ -937,6 +957,7 @@ object CelebornConf extends Logging {
       .version("0.2.0")
       .doc("Max chunk size of reducer's merged shuffle data. For example, if a reducer's " +
         "shuffle data is 128M and the data will need 16 fetch chunk requests to fetch.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:960]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 18]
```diff
@@ -946,6 +967,7 @@ object CelebornConf extends Logging {
       .categories("client")
       .version("0.2.0")
       .doc("Max retry times for client to register shuffle.")
+      .version("0.2.0")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:970]: duplicated version assignment

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 6 above)
Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1803]: the doc says
> If direct memory usage reach this limit ...

Is it dedicated to direct memory? If yes, should we mention it in the name?

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 7 above)
Comment (nafiyAix) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:867]: Get it, will keep the order like doc().version()

Comment (nafiyAix) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:867]: Seems CelebornConf's methods calling order is not consistent when building ConfigEntry of various config. I will delete duplicated version in this pr, maybe we can sort out all builder's calling order in other pr.

Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:867]: yea, either order is fine, as long as consistent

**Review by pan3793 — COMMENTED** [review]
My proposed names, any ideas are welcome
```
celeborn.worker.partitionSorter.directMemoryRatioThreshold
celeborn.worker.directMemoryRatioToPauseReceive
celeborn.worker.directMemoryRatioToPauseReplicate
celeborn.worker.directMemoryRatioToResume
celeborn.worker.memory.checkInterval
celeborn.worker.memory.reportInterval
```

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 19]
```diff
@@ -1754,32 +1764,66 @@ object CelebornConf extends Logging {
...
+      .createWithDefault(0.95)
 
-  def workerResumeRatio(conf: CelebornConf): Double = {
-    conf.getDouble("rss.resume.memory.ratio", 0.5)
-  }
+  val WORKER_RESUME_DIRECT_MEMORY_RATIO: ConfigEntry[Double] =
+    buildConf("celeborn.worker.resume.directMemory.ratio")
+      .withAlternative("rss.resume.memory.ratio")
+      .categories("worker")
+      .doc("If direct memory usage is less than this  limit, worker will resume receive.")
```
Comment (pan3793) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1801]: it's actually to resume all works, not only for "receive"

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 20]
```diff
@@ -1754,32 +1764,66 @@ object CelebornConf extends Logging {
...
+      .createWithDefaultString("10ms")
 
-  def workerDirectMemoryPressureCheckIntervalMs(conf: CelebornConf): Int = {
-    conf.getInt("rss.worker.memory.check.interval", 10)
-  }
+  val WORKER_DIRECT_MEMORY_REPORT_INTERVAL: ConfigEntry[Long] =
+    buildConf("celeborn.worker.memory.report.interval")
+      .withAlternative("rss.worker.memory.report.interval")
+      .categories("worker")
+      .doc("Worker direct memory tracker report interval, its timeunit is second.")
```
[the remark this answers is not in the document]
Comment (nafiyAix) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1820]: I think it only report as loginfo, so do we need to change its name?
image

[the remark this answers is not in the document]
Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1820]: The name is fine, but we should clarify the behavior in docs.

[the remark this answers is not in the document]
Comment (nafiyAix) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1820]: BTW, for the `rss.worker.memory.check.interval` i change its default value format from `createWithDefault(10)` to  `createWithDefaultString("10ms")` in order to keep consistency with other time config, may i know is there a way to tolerate 2 types of default value?

[the remark this answers is not in the document]
Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1820]: I should work when the unit is eliminated.
```
 /**
   * Convert a passed time string (e.g. 50s, 100ms, or 250us) to a time count in the given unit. The
   * unit is also considered the default if the given string does not specify a unit.
   */
  public static long timeStringAs(String str, TimeUnit unit) {
```

[the remark this answers is not in the document]
Comment (nafiyAix) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1820]: Get it thanks

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 21]
```diff
@@ -1754,32 +1765,66 @@ object CelebornConf extends Logging {
...
+      .createWithDefaultString("10ms")
 
-  def workerDirectMemoryPressureCheckIntervalMs(conf: CelebornConf): Int = {
-    conf.getInt("rss.worker.memory.check.interval", 10)
-  }
+  val WORKER_DIRECT_MEMORY_REPORT_INTERVAL: ConfigEntry[Long] =
+    buildConf("celeborn.worker.memory.reportInterval")
+      .withAlternative("rss.worker.memory.report.interval")
+      .categories("worker")
+      .doc("Worker direct memory tracker reporting to log interval.")
```
Comment (AngersZhuuuu) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1821]: `Worker direct memory tracker reporting interval.`

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 22]
```diff
@@ -1754,32 +1765,66 @@ object CelebornConf extends Logging {
...
+      .createWithDefault(0.85)
 
-  def workerPauseRepcaliteRatio(conf: CelebornConf): Double = {
-    conf.getDouble("rss.pause.replicate.memory.ratio", 0.95)
-  }
+  val WORKER_DIRECT_MEMORY_RATIO_PAUSE_REPLICATE: ConfigEntry[Double] =
+    buildConf("celeborn.worker.directMemoryRatioToPauseReplicate")
+      .withAlternative("rss.pause.replicate.memory.ratio")
+      .categories("worker")
+      .doc("If direct memory usage reaches this limit, worker will stop to receive from executor and other worker.")
```
Comment (AngersZhuuuu) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1792]: `If direct memory usage reaches this limit, the worker will stop to receive replication data from executor and other workers.`

**Review by AngersZhuuuu — APPROVED** [review]
LGTM  except some configuration docs

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 23]
```diff
@@ -1754,32 +1765,66 @@ object CelebornConf extends Logging {
...
   }
 
-  def partitionSortMaxMemoryRatio(conf: CelebornConf): Double = {
-    conf.getDouble("rss.partition.sort.memory.max.ratio", 0.1)
-  }
+  val PARTITION_SORTER_DIRECT_MEMORY_RATIO_THRESHOLD: ConfigEntry[Double] =
+    buildConf("celeborn.worker.partitionSorter.directMemoryRatioThreshold")
+      .withAlternative("rss.partition.sort.memory.max.ratio")
+      .categories("worker")
+      .doc("Max ratio of sort memory.")
```
Comment (AngersZhuuuu) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1772]: More detail? what's the behavior when read the threshold?

**Review by pan3793 — APPROVED** [review]

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 21 above)
Comment (nafiyAix) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1821]: Have discussed with pan cheng about this doc before, https://github.com/alibaba/celeborn/pull/831#discussion_r1004211709 ,but current doc seems also a bit weird, will change it.

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 22 above)
Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1792]: > ... to receive replication data ...

Should we emphasize "replication" here?

**Pull request merged**